### PR TITLE
Add junit and metadata subcommands to kntest

### LINKED ISSUE #1871: Text under DataRaceAnalysis tab of Spyglass log page does not have correct line feeds
See https://prow.knative.dev/view/gcs/knative-prow/logs/ci-knative-serving-continuous/1232488190519021569

/bug
/kind good-first-issue

### PR DESCRIPTION
**Pull request by chizhg**
**What this PR does, why we need it**:
1. Some cleanups to the cluster subcommands
2. Add junithelper and metahelper as subcommands of `kntest`, later then we can delete them from https://github.com/knative/pkg/tree/master/testutils
3. Replace one function in the bash scripts with `kntest junit` subcommand, this fixes https://github.com/knative/test-infra/issues/1871

**Which issue(s) this PR fixes**:
Fixes #1871

### CODE REVIEW

File: kntest/pkg/cluster/gke/README.md [hunk 1]
```diff
@@ -9,28 +9,28 @@ cluster
 
 ## Usage
 
-This tool can be invoked from command line with following parameters:
+This tool can be invoked from command line.
+The following parameters are common for all subcommands:
```
Comment (mattmoor) [inline kntest/pkg/cluster/gke/README.md:13]: Format markdown:
```suggestion
This tool can be invoked from command line. The following parameters are common
for all subcommands:
```

File: kntest/pkg/cluster/gke/README.md [hunk 2]
```diff
@@ -9,28 +9,28 @@ cluster
...
+- `--region`: GKE region, default "us-central1". \
+  Can be more than one to set as backup regions.
+- `--resource-type`: Boskos resource type, default "gke-project"
 
 ## Subcommands
 
 ### Create
 
 `kntest cluster gke create` will always create a new cluster.
+It accepts the following extra parameters:
```
Comment (mattmoor) [inline kntest/pkg/cluster/gke/README.md:26]: Format markdown:
```suggestion
`kntest cluster gke create` will always create a new cluster. It accepts the
following extra parameters:
```

File: kntest/pkg/metadata/README.md [hunk 3]
```diff
@@ -0,0 +1,32 @@
+## kntest metadata
+
+`kntest metadata` command is used for operations on the metadata.json file
+in Prow job artifacts.
```
Comment (mattmoor) [inline kntest/pkg/metadata/README.md:4]: Remove trailing whitespace:
```suggestion
in Prow job artifacts.
```

File: kntest/pkg/metadata/README.md [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+This tool can be invoked from command line with following parameters:
+
+- `--suite`: name of suite
+- `--name`: name of test
+- `--err-msg`: (optional) error message, by default it's empty, means test
+  passed
+- `--dest`: (optional) file path for result to be written to, default
+  `junit_result.xml`
+
```
Comment (mattmoor) [inline kntest/pkg/metadata/README.md:16]: Remove trailing whitespace:
```suggestion

```

File: kntest/pkg/metadata/README.md [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
```
Comment (mattmoor) [inline kntest/pkg/metadata/README.md:16]: Remove trailing whitespace:
```suggestion

```

File: kntest/pkg/junit/README.md [hunk 6]
```diff
@@ -0,0 +1,22 @@
+## kntest junit
+
+`kntest junit` command is used for creating a simple junit test result, which is used for
+creating a single junit result file with a single test.
```
Comment (mattmoor) [inline kntest/pkg/junit/README.md:4]: Format markdown:
```suggestion
`kntest junit` command is used for creating a simple junit test result, which is
used for creating a single junit result file with a single test.
```

File: kntest/pkg/metadata/README.md [hunk 7]
```diff
@@ -0,0 +1,32 @@
+## kntest metadata
+
+`kntest metadata` command is used for operations on the metadata.json file
+in Prow job artifacts.
```
Comment (mattmoor) [inline kntest/pkg/metadata/README.md:4]: Format markdown:
```suggestion
`kntest metadata` command is used for operations on the metadata.json file in
Prow job artifacts.
```

File: kntest/pkg/metadata/README.md (hunk 5 above)
Comment (mattmoor) [inline kntest/pkg/metadata/README.md:16]: Format markdown:
```suggestion

```

File: kntest/pkg/metadata/README.md [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+## Subcommands
+
+### Set
+
+`kntest metadata set` subcommand can be invoked with following parameters:
+
+- `--key`: meta info key
+- `--value`: meta info value
+
+Note the value can be overwritten if the key already exists in the metadata.json file.
```
Comment (mattmoor) [inline kntest/pkg/metadata/README.md:26]: Format markdown:
```suggestion
Note the value can be overwritten if the key already exists in the metadata.json
file.
```

File: kntest/pkg/junit/README.md (hunk 6 above)
Comment (mattmoor) [inline kntest/pkg/junit/README.md:4]: Format markdown:
```suggestion
`kntest junit` command is used for creating a simple junit test result, which is
used for creating a single junit result file with a single test.
```

File: kntest/pkg/metadata/README.md (hunk 7 above)
Comment (mattmoor) [inline kntest/pkg/metadata/README.md:4]: Format markdown:
```suggestion
`kntest metadata` command is used for operations on the metadata.json file in
Prow job artifacts.
```

File: kntest/pkg/metadata/README.md (hunk 5 above)
Comment (mattmoor) [inline kntest/pkg/metadata/README.md:16]: Format markdown:
```suggestion

```

File: kntest/pkg/metadata/README.md (hunk 8 above)
Comment (mattmoor) [inline kntest/pkg/metadata/README.md:26]: Format markdown:
```suggestion
Note the value can be overwritten if the key already exists in the metadata.json
file.
```

File: kntest/pkg/metadata/README.md [hunk 9]
```diff
@@ -0,0 +1,32 @@
+## kntest metadata
+
+`kntest metadata` command is used for operations on the metadata.json file in
+Prow job artifacts.
```
Comment (mattmoor) [inline kntest/pkg/metadata/README.md:4]: Format markdown:
```suggestion
`kntest metadata` command is used for operations on the metadata.json file in
Prow job artifacts.
```

File: kntest/pkg/metadata/README.md [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+## Usage
+
+This tool can be invoked from command line with following parameters:
+
+- `--suite`: name of suite
+- `--name`: name of test
+- `--err-msg`: (optional) error message, by default it's empty, means test
+  passed
+- `--dest`: (optional) file path for result to be written to, default
+  `junit_result.xml`
```
Comment (mattmoor) [inline kntest/pkg/metadata/README.md:15]: Format markdown:
```suggestion
  `junit_result.xml`

```

File: kntest/pkg/metadata/README.md [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+## Subcommands
+
+### Set
+
+`kntest metadata set` subcommand can be invoked with following parameters:
+
+- `--key`: meta info key
+- `--value`: meta info value
+
+Note the value can be overwritten if the key already exists in the metadata.json
```
Comment (mattmoor) [inline kntest/pkg/metadata/README.md:25]: Format markdown:
```suggestion
Note the value can be overwritten if the key already exists in the metadata.json
file.
```

File: kntest/pkg/metadata/README.md (hunk 10 above)
Comment (mattmoor) [inline kntest/pkg/metadata/README.md:15]: Format markdown:
```suggestion
  `junit_result.xml`

```

File: kntest/pkg/metadata/README.md (hunk 11 above)
Comment (mattmoor) [inline kntest/pkg/metadata/README.md:25]: Format markdown:
```suggestion
Note the value can be overwritten if the key already exists in the metadata.json
file.
```

File: kntest/pkg/metadata/README.md (hunk 10 above)
Comment (mattmoor) [inline kntest/pkg/metadata/README.md:15]: Format markdown:
```suggestion
  `junit_result.xml`

```

**Comment by chizhg** [comment]
/hold
Waiting for https://github.com/knative/pkg/pull/1278 to be merged and then rebase.

**Comment by chizhg** [comment]
/hold cancel
/cc @chaodaiG 

/cc @albertomilan 
For some visibility on the work the team is doing.

File: kntest/pkg/junit/command.go [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+			}
+			suite.AddTestCase(tc)
+			// Ignore the error as it only happens if the test suite name already exists.
+			suites.AddTestSuite(&suite)
+			contents, err := suites.ToBytes("", "")
+			if err != nil {
+				log.Fatal(err)
+			}
+			if err := ioutil.WriteFile(dest, contents, 0644); err != nil {
+				log.Fatalf("Error writing to file %q: %v", dest, err)
```
Comment (chaodaiG) [inline kntest/pkg/junit/command.go:62]: error message start with lower case

File: kntest/cmd/kntest/main.go [hunk 13]
```diff
@@ -36,8 +38,10 @@ func main() {
 	}
 
 	cluster.AddCommands(cmds)
+	junit.AddCommands(cmds)
+	metadata.AddCommands(cmds)
 
 	if err := cmds.Execute(); err != nil {
-		log.Fatalf("error during command execution: %v", err)
+		log.Fatalf("Error during command execution: %v", err)
```
Comment (chaodaiG) [inline kntest/cmd/kntest/main.go:45]: Why capitalize the first letter?

File: kntest/pkg/metadata/command.go [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+				log.Fatalf("error setting meta info for %q=%q: %v", key, value, err)
+			}
+		},
+	}
+	setCmd.Flags().StringVar(&value, "value", "", "meta info value")
+	metadataCmd.AddCommand(setCmd)
+}
+
+func addGetCommand(metadataCmd *cobra.Command, c *client.Client) {
+	var setCmd = &cobra.Command{
```
[the remark this answers is not in the document]
Comment (chizhg) [reply kntest/pkg/metadata/command.go]: Done.

File: kntest/pkg/junit/command.go [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+import (
+	"io/ioutil"
+	"log"
+
+	"github.com/spf13/cobra"
+	"knative.dev/pkg/test/junit"
+)
+
+var (
+	suite  string
```
[the remark this answers is not in the document]
Comment (chizhg) [reply kntest/pkg/junit/command.go:28]: Done.

File: kntest/pkg/junit/command.go (hunk 12 above)
Comment (chizhg) [reply kntest/pkg/junit/command.go]: From https://github.com/golang/go/wiki/CodeReviewComments#error-strings, `fmt.Errorf("Something bad")` is bad but it's recommended to have logging capitalized.

File: kntest/cmd/kntest/main.go (hunk 13 above)
Comment (chizhg) [reply kntest/cmd/kntest/main.go:45]: Ditto

File: kntest/pkg/junit/command.go (hunk 12 above)
Comment (chaodaiG) [reply kntest/pkg/junit/command.go]: gotcha, good to learn

**Pull request merged**